Feature/events index

borourke opened the pull request:
Hey @cchanningallen , happy to announce the first PR here!

Gives events_index the payload we discussed in the [SwaggerHub document](https://app.swaggerhub.com/api/bryan.orourke24/Circles/1.0.0)

I'm doing TDD by defining an API Payload for the endpoint in a spec, then developing against that.  This will be the first project I'm not writing unit tests, but it's becoming more accepted that unit tests don't hold much value in this scenario (who cares which methods are defined as long as the payload returned is expected)

borourke commented on app/controllers/api/events_controller.rb:
```diff
@@ -0,0 +1,6 @@
+class API::EventsController < ApplicationController
+  def index
```
This will be a typical controller.  So let me know if you liked a different approach.

cchanningallen replied:
If we validate / filter `params`, could we just pass `events = EventIconBuilder.new(params)`?

borourke replied:
Great point

borourke commented on app/models/event.rb:
```diff
@@ -0,0 +1,22 @@
+  has_many :event_members
+  has_many :users, through: :event_members
+
+  def to_icon(event_number:)
```
I decided against using some more heavyweight solutions like activerecord serialization because we'll want to customize every payload anyways.  so I figured it's nice to just be more explicit and define the full custom payload as opposed to having you look back at the schema to see what else is going into the payload

cchanningallen replied:
Works for me

borourke commented on app/models/event.rb:
```diff
@@ -0,0 +1,22 @@
+    }
+  end
+
+  def add_members(members:)
```
Right now this is only used in specs.  But I'm thinking I would keep it in the model because at some point we'll need this.

cchanningallen replied:
Sounds good

borourke commented on lib/event_icon_builder.rb:
```diff
@@ -0,0 +1,36 @@
+class EventIconBuilder
```
A lot of the stuff in this class will be moved out to an abstract class that this and other "Builders" will inherit from.  But since we only have one, for right now, I figured not to over engineer it.  Once we get a second route/object going, I'll make this more generalizable.

cchanningallen replied:
Got it, sounds good

cchanningallen replied:
A thought about the class name - this is building a payload of event icon**s** (plural), so does the `EventIconBuilder` (singular) class name make sense?

borourke replied:
Raise a good point. I'll rename it I be plural

cchanningallen replied:
I remember looking at [`JBuilder`](https://github.com/rails/jbuilder) a while ago and thinking it looked neat

borourke replied:
Yeah, I know I'm in the minority on this, but I always end up thinking it's extra dressing where all you need is a json blob in a controller. Most people put it in the view and use syntax helpers, which I'm not opposed to, but this has always felt simpler. And when rails5 API mode took away the views folder I felt emboldened

borourke replied:
But I definitely agree that this could be cleaned up structure wise to more closely resemble a json view folder structure. Will give this another pass.

cchanningallen commented on app/models/event.rb:
```diff
@@ -0,0 +1,22 @@
+      event_number: event_number,
+      start_date: start_date,
+      end_date: end_date,
+      image_url: image_url,
```
so funny to me that all this duplication is necessary. In ES6, it'd just be 

```
{ id, event_number, ... }
```

borourke replied:
Hahaha yeah ruby is pretty bangin but can be a bit boiler plate at times

cchanningallen commented on lib/event_icon_builder.rb:
```diff
@@ -0,0 +1,36 @@
+      # The index + start_id means if you ask for start_id 30, then the first event in 
+      # the returned payload will be 30.  It counts upwards from there to stop at
+      # start_id + count.
+      event.to_icon(event_number: index + start_id)
```
Hmm - only thing that seems like a bit wonky is that the view logic (i.e., json payload generator) seems to be housed in the model for an individual event, but housed here for the entire events payload. Might it make more sense to have an `EventIconsBuilder` (plural) composed of `EventIconBuilder` (singular)?

Also, ideally all data coming over the wire would be in camelCase. Otherwise, I'll probably end up writing adapters on the client to convert it. No biggie, but if we're already coercing the raw data into a payload here, might as well translate it for final consumption to save me coding time on the client.

borourke replied:
Noted on the camel case, no biggie to change to that. And I like your idea about having the plural model (the payload) be composed if singular models.

Also makes it easier to generalize I think.

cchanningallen commented on spec/controllers/api/events_controller_spec.rb:
```diff
@@ -0,0 +1,54 @@
+        expect(response.body).to eq(expected_payload)
+      end
+    end
+
```
Looks good! One thought is that it doesn't seem we test for individual attributes in the return data; if, for example, we're expecting a `name` attribute on the first `EventIcon`, but the `EventIconBuilder` was updated and now isn't returning the required data, I don't think this test would detect it.

borourke replied:
Yeah I should really test the whole structure. Good catch

cchanningallen commented:
Looks so damn good overall! Let me know if my comments about where to build the payload & testing don't make sense; it's been a while since I've been in the back-end of the woods, after all :)

borourke commented:
I actually don't see any of the comments? Am I missing them or did you forget to press enter or something?

borourke commented:
Great feedback, really thoughtful thank you!

I will give this another pass tomorrow and ping you again

cchanningallen commented:
Word! Like I said, awesome overall. Front-end code is almost done, so we should have something to look at tomorrow / this weekend. Let's plan another call for Sat or Sun?

cchanningallen commented:
Getting there on my end; your code works great :)

<img width="1440" alt="screen shot 2016-11-11 at 7 06 05 am" src="https://cloud.githubusercontent.com/assets/2758435/20217497/69071c7e-a7dd-11e6-9e2f-acaf7c8fd776.png">

borourke commented:
Glad to hear :) This weekend I'm going to the city to visit family and for a friends bday, but next week anyday before Thursday works for me?

borourke commented:
That's a real UI too which is pretty cool!  That alone would have taken me hours and hours in erb

borourke commented on app/serializers/event_serializer.rb:
```diff
@@ -0,0 +1,6 @@
+class EventSerializer < ActiveModel::Serializer
```
Before I move onto making the specs more robust what do you think about this set up?  I decided to give ActiveRecord Serialization a try.  Definitely makes small files.  Down side is I can't really customize which fields we take from a model for each payload.

borourke replied:
Here's an example of a return payload:

```
curl -X GET http://localhost:3000/api/events?user_id=1
[{"id":1,"name":"Thanksgiving","image_url":null,"start_date":"2016-11-28","end_date":null,"users":[{"id":1,"name":"Channing","image_url":null},{"id":2,"name":"Bryan","image_url":null},{"id":3,"name":"Eileen","image_url":null}],"event_members":[{"id":1,"user_id":1,"event_id":1,"created_at":"2016-11-08T00:22:20.335Z","updated_at":"2016-11-08T00:22:20.335Z"},{"id":2,"user_id":2,"event_id":1,"created_at":"2016-11-08T00:22:20.339Z","updated_at":"2016-11-08T00:22:20.339Z"},{"id":3,"user_id":3,"event_id":1,"created_at":"2016-11-08T00:22:20.342Z","updated_at":"2016-11-08T00:22:20.342Z"}]},{"id":2,"name":"Boulder Visit","image_url":null,"start_date":"2016-12-09","end_date":"2016-12-13","users":[{"id":1,"name":"Channing","image_url":null},{"id":2,"name":"Bryan","image_url":null},{"id":3,"name":"Eileen","image_url":null}],"event_members":[{"id":4,"user_id":1,"event_id":2,"created_at":"2016-11-08T00:22:20.351Z","updated_at":"2016-11-08T00:22:20.351Z"},{"id":5,"user_id":2,"event_id":2,"created_at":"2016-11-08T00:22:20.353Z","updated_at":"2016-11-08T00:22:20.353Z"},{"id":6,"user_id":3,"event_id":2,"created_at":"2016-11-08T00:22:20.355Z","updated_at":"2016-11-08T00:22:20.355Z"}]}]
```

borourke replied:
Also another downside is the fields need to match the active record model, meaning I can't convert the keys to camelCase.  Summary is: really simple, not much code, but not much customization allowed either.

cchanningallen replied:
Yikes. I'm ok with the `camelCase`, but to optimize our payloads we're going to want the ability to send specific data from models. Seems like [jBuilder](https://github.com/rails/jbuilder) might be our best bet.

borourke replied:
Yeah I agree.  The flexibility is a must.  I'll change this to jBuilder tonight.

cchanningallen commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,11 @@
+
+  def count
+    params[:count] || default_count
+  end
```
why are these methods on ApplicationController? seems like they're only relevant for the events index

borourke replied:
That's true for right now.  But I'm thinking pagination will be a frequent feature of our application.  And most controllers will need this.  If it's confusing for right now though happy to move it back.  It's a trivial change.

borourke commented on app/controllers/application_controller.rb:
```diff
@@ -1,2 +1,37 @@
+    render json: { error: exception }, status: 403
+  end
+
+  # The follow section of the ApplicationController is for checking
```
@cchanningallen added a params check to make sure we have intelligent responses if a required param is missing.  Output looks like this:

```
{"error":"The following param(s) are required but were not found: user_id"}

borourke commented on spec/controllers/api/events_controller_spec.rb:
```diff
@@ -0,0 +1,75 @@
+        end
+      end
+
+      it 'returns the expected payload' do
```
Specs now check for stats code, keys in the responses and number of items in the response.

borourke commented on app/views/api/events/index.json.jbuilder:
```diff
@@ -0,0 +1,13 @@
+json.events events do |event|
```
Here is the jbuilder code.  It's nice to work with, definitely better than active record serialization.  Here is a sample of this response (note the camelCase :)):

```
{"events":[{"id":1,"name":"Thanksgiving","startDate":"2016-11-28","endDate":null,"imageUrl":null,"members":[{"id":1,"name":"Channing","imageUrl":null},{"id":2,"name":"Bryan","imageUrl":null},{"id":3,"name":"Eileen","imageUrl":null}]},{"id":2,"name":"Boulder Visit","startDate":"2016-12-09","endDate":"2016-12-13","imageUrl":null,"members":[{"id":1,"name":"Channing","imageUrl":null},{"id":2,"name":"Bryan","imageUrl":null},{"id":3,"name":"Eileen","imageUrl":null}]}]}

borourke commented:
Alrighty, after experimenting with several JSON building versions, I think this is a good one.  Also note I added the params requirement check if you want to look that over real quick and fixed the specs to be less self-fulfilling.

cchanningallen commented:
Looks awesome! Well done brother man :)

cchanningallen commented on config/environment.rb:
```diff
@@ -3,3 +3,7 @@
+
+# Sets JBuilder to use camelCase keys
+# in the JSON by default
+Jbuilder.key_format camelize: :lower
```
💯
